File: HeadsUpWindow.rb
require 'osx/cocoa'

class HeadsUpWindow < OSX::NSWindow
	def	initWithLocation(location)
		initWithContentRect_styleMask_backing_defer(frame_for_location(location), OSX::NSBorderlessWindowMask, OSX::NSBackingStoreBuffered, false)
		
		@location = location
		
		setAutodisplay(true)
    setBackgroundColor(OSX::NSColor.clearColor)
		setContentView(HeadsUpTextView.alloc.initWithFrame(frame))
		setHasShadow(false)
		setIgnoresMouseEvents(true)
		setLevel(OSX::KCGDesktopWindowLevel)
		setOpaque(false)
		setReleasedWhenClosed(true)
		
		orderFront(self)
		self
	end
	
	def update_contents(string)
	  contentView.setString(bottom_justify(string))
	end
	
	private
	
	def bottom_justify(string)
    lines = string.split(/\n/)
    (12 - lines.size).times { lines.unshift('') }
    lines.join("\n")
  end

	def frame_for_location(location)
	  main_frame = OSX::NSScreen.mainScreen.frame
	
	  margin = 12
	  width  = 600
	  height = 192
	
	  case location
    when :bottom_left
	    [margin, margin, width, height]
	  when :bottom_right
	    [main_frame.width - width - margin, margin, width, height]
    else
      raise "Unsupported window location: #{location}"
    end
	end
end

File: HeadsUpWindow.m
#import "HeadsUpTextView.h"
#import "HeadsUpWindow.h"

@implementation HeadsUpWindow

- (id)initWithPosition:(WindowPosition *)thePosition {
	self = [super initWithContentRect:[thePosition windowFrameWithSize:NSMakeSize(100, 100)] styleMask:NSBorderlessWindowMask backing:NSBackingStoreBuffered defer:FALSE];

	if (self) {
		position = thePosition;

		[self setBackgroundColor: [NSColor clearColor]];
		[self setContentView:[[HeadsUpTextView alloc] init]];
		[self setIgnoresMouseEvents:TRUE];
		[self setLevel:CGWindowLevelForKey(kCGDesktopWindowLevelKey)];
		[self setOpaque:FALSE];
	}
	
	return self;
}

- (void)headsUpScreenDidUpdate:(NSNotification *)notification {
	HeadsUpScreen *screen = [notification object];
	[[self contentView] setString:[screen contents]];
	[self setFrame:[position windowFrameWithSize:[(HeadsUpTextView *) [self contentView] textSize]] display:TRUE];
}

@end